### PR TITLE
Specific callbacks port added

### PR DESCRIPTION
**Pull request by yescorp**

### CODE REVIEW

File: tool/src/callback_dictionary.ts [hunk 1]
```diff
@@ -1,28 +1,130 @@
...
+    typescript(hljs),
   ].map(language => getLanguageCallbacks(language))),
 ]);
 
 function getLanguageCallbacks(language: StringObject<any>): [string, string][] {
-  // TODO: Extract all callbacks recursively.
-  return [["a", "b"]];
+  let result: [string, string][] = [];
+
+  const addResult = (k: string, v: string) => {
```
Comment (alexeyinkin) [inline tool/src/callback_dictionary.ts:43]: Non-lambda.

File: tool/src/callback_dictionary.ts [hunk 2]
```diff
@@ -1,28 +1,130 @@
...
+    if (!commonCallbacks.has(v.toString())) {
```
Comment (alexeyinkin) [inline tool/src/callback_dictionary.ts:44]: Add the language name to the key.

File: tool/src/callback_dictionary.ts [hunk 3]
```diff
@@ -1,28 +1,130 @@
...
+    php(hljs),
+    ruby(hljs),
+    typescript(hljs),
   ].map(language => getLanguageCallbacks(language))),
 ]);
 
 function getLanguageCallbacks(language: StringObject<any>): [string, string][] {
-  // TODO: Extract all callbacks recursively.
-  return [["a", "b"]];
+  let result: [string, string][] = [];
```
Comment (alexeyinkin) [inline tool/src/callback_dictionary.ts:41]:
```suggestion
  const result: [string, string][] = [];
```

File: tool/src/callback_dictionary.ts [hunk 4]
```diff
@@ -1,28 +1,130 @@
...
+
+  const addResult = (k: string, v: string) => {
+    if (!commonCallbacks.has(v.toString())) {
+      result.push([v.toString(), k]);
+    }
+  };
+
+  let objects = <StringObject<any>>[];
+  let languageName: string = language.name;
+  let currentName: string[] = ['k'];
```
Comment (alexeyinkin) [inline tool/src/callback_dictionary.ts:51]:
```suggestion
  let currentName: string[] = ['language'];
```

File: tool/src/callback_dictionary.ts [hunk 5]
```diff
@@ -1,28 +1,130 @@
...
+  if (languageName !== undefined) {
+    currentName.push(languageName.toLowerCase());
```
Comment (alexeyinkin) [inline tool/src/callback_dictionary.ts:53]: Throw an exception if there is no name, and abort the entire porting.

File: tool/src/callback_dictionary.ts [hunk 6]
```diff
@@ -1,28 +1,130 @@
...
+  }
+
+  function iterate(language: StringObject<any>) {
```
Comment (alexeyinkin) [inline tool/src/callback_dictionary.ts:56]:
```suggestion
  function iterate(language: Mode): void {
```

File: tool/src/callback_dictionary.ts [hunk 7]
```diff
@@ -1,28 +1,130 @@
...
-  return [["a", "b"]];
+  let result: [string, string][] = [];
+
+  const addResult = (k: string, v: string) => {
+    if (!commonCallbacks.has(v.toString())) {
+      result.push([v.toString(), k]);
+    }
+  };
+
+  let objects = <StringObject<any>>[];
```
Comment (alexeyinkin) [inline tool/src/callback_dictionary.ts:49]:
```suggestion
  const visitedModes = new Set<StringObject<any>>();
```
?

File: tool/src/callback_dictionary.ts [hunk 8]
```diff
@@ -1,28 +1,130 @@
...
+  let currentName: string[] = ['k'];
+  if (languageName !== undefined) {
+    currentName.push(languageName.toLowerCase());
+  }
+
+  function iterate(language: StringObject<any>) {
+    if (objects.includes(language)) {
+      return;
+    }
+    else {
```
Comment (alexeyinkin) [inline tool/src/callback_dictionary.ts:60]: No else after return.

File: tool/src/callback_dictionary.ts [hunk 9]
```diff
@@ -1,28 +1,130 @@
...
+      objects.push(language);
+    }
+
+    if (language === null) {
+      return;
+    }
```
Comment (alexeyinkin) [inline tool/src/callback_dictionary.ts:66]: Try to get rid of `null` with a compile-time check.

File: tool/src/callback_dictionary.ts [hunk 10]
```diff
@@ -1,28 +1,130 @@
...
+
+    let entries = Object.entries(language);
+
+    for (const item of entries) {
+      let [k, v] = item;
+
+      switch (k) {
+        case "on:begin":
+          currentName.push('onBegin');
+          const onBeginCallback = currentName.reduce((prev, current) => {
```
Comment (alexeyinkin) [inline tool/src/callback_dictionary.ts:76]:
```suggestion
          addResult(
            [...currentName, 'onBegin'].join('_'),
            v,
          );
```

File: tool/src/callback_dictionary.ts [hunk 11]
```diff
@@ -1,28 +1,130 @@
...
+          const onEndCallback = currentName.reduce((prev, current) => {
+            return prev + "_" + current;
+          });
+          addResult(onEndCallback, v);
+          currentName.pop();
+          break;
+
+        case "starts":
+          currentName.push(k);
+          iterate(v);
```
Comment (alexeyinkin) [inline tool/src/callback_dictionary.ts:94]: Check if `v` is not an object. Then throw an exception and break the entire porting.
This allows to remove `null` check above.

File: tool/src/specific_ports.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+export async function portMathematicaSymbolsSet(): Promise<any> {
+    const notice_comment = "// GENERATED CODE - DO NOT MODIFY BY HAND\n";
```
Comment (alexeyinkin) [inline tool/src/specific_ports.ts:2]: Create or reuse a global constant.

File: tool/js/highlight.js [hunk 13]
```diff
@@ -6,6 +6,7 @@ import path from "path";
 
 import { callbackDictionary } from "./callback_dictionary.js";
 import { expandRefs, getLodashGetKey } from './porting.js';
+import { portMathematicaSymbolsSet } from './specific_ports.js';
```
Comment (alexeyinkin) [inline tool/js/highlight.js:9]:
```suggestion
import { portMathematicaSpecific } from './specific_ports/mathematica.js';
```

**Review by alexeyinkin — CHANGES_REQUESTED** [review]

File: tool/src/callback_dictionary.ts (hunk 5 above)
Comment (yescorp) [reply tool/src/callback_dictionary.ts:53]: Turns out, PHP doesn't have a `name` field...

It does have a callback on the main branch, but not in the latest version.

File: tool/src/callback_dictionary.ts (hunk 6 above)
Comment (yescorp) [reply tool/src/callback_dictionary.ts:56]: If I remove the placeholder `d.ts` file, I get the following message.
Investigating this issue.

Could not find a declaration file for module 'highlight.js'. '/Users/yestay/work/Opensource/dart-highlights/tool/node_modules/highlight.js/es/index.js' implicitly has an 'any' type.
  There are types at '/Users/yestay/work/Opensource/dart-highlights/tool/node_modules/highlight.js/types/index.d.ts', but this result could not be resolved when respecting package.json "exports". The 'highlight.js' library may need to update its package.json or typings.ts(7016)

File: tool/src/callback_dictionary.ts (hunk 2 above)
Comment (yescorp) [reply tool/src/callback_dictionary.ts:42]: The problem here is that the key is the source code of the callback that we match when replacing it with the function name.

I couldn't find a way how we can casually add language name to the key.
My suggestion is to leave everything as is, and if there are new callbacks in `highlight.js`, we would need to make manual changes anyway. So, I think it is not that big of a deal.

**Review by alexeyinkin — APPROVED** [review]

**Pull request merged**